File: secure-kyc-chain/src/types/kyc.ts
export type UserRole = 'user' | 'reviewer' | 'admin' | 'institution';

export type KYCStatus = 
  | 'DRAFT' 
  | 'REGISTERED'
  | 'UPLOADED' 
  | 'PROCESSING'
  | 'IN_REVIEW' 
  | 'VERIFIED'  // UKN issued
  | 'APPROVED'  // Legacy - use VERIFIED
  | 'REJECTED'
  | 'SUSPENDED'
  | 'EXPIRED'
  | 'REQUEST_INFO';

export type DocumentType = 
  | 'PASSPORT'
  | 'DRIVERS_LICENSE'
  | 'NATIONAL_ID'
  | 'SELFIE';

export interface User {
  id: string;
  email: string;
  role: UserRole;
  created_at: string;
}

export interface ExtractedData {
  name?: string;
  dob?: string;
  address?: string;
  documentNumber?: string;
  expiryDate?: string;
  nationality?: string;
}

export interface Document {
  id: string;
  kyc_id: string;
  doc_type: DocumentType;
  file_path: string;
  file_hash: string;
  extracted_data: ExtractedData;
  verified: boolean;
  uploaded_at: string;
}

export interface ShapFeature {
  feature: string;
  shap_value: number;
  feature_value: string | number;
}

export interface Verification {
  id: string;
  kyc_id: string;
  event_type: string;
  details: Record<string, any>;
  performed_by: string;
  timestamp: string;
  tx_hash?: string;
}

export interface KYCApplication {
  id: string;
  user_id: string;
  user_email?: string;
  ukn?: string;  // Unique KYC Number
  status: KYCStatus;
  risk_score: number | null;
  shap_explanation?: ShapFeature[];
  created_at: string;
  updated_at: string;
  documents: Document[];
  verifications: Verification[];
  face_match_score?: number;
  reviewer_comment?: string;
  blockchain_tx_hash?: string;
  verified_at?: string;
  expires_at?: string;
}

export interface AuditRecord {
  id: string;
  entity_type: string;
  entity_id: string;
  event_hash: string;
  tx_hash: string;
  timestamp: string;
  details?: Record<string, any>;
}

export interface RiskMetrics {
  total_applications: number;
  auto_approved: number;
  manual_reviews: number;
  rejected: number;
  avg_risk_score: number;
  avg_processing_time: number;
}

File: secure-kyc-chain/backend/app/schemas/kyc.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class ExtractedData(BaseModel):
    name: Optional[str] = None
    dob: Optional[str] = None
    address: Optional[str] = None
    documentNumber: Optional[str] = None
    expiryDate: Optional[str] = None
    nationality: Optional[str] = None


class ShapFeature(BaseModel):
    feature: str
    shap_value: float
    feature_value: str | float


class DocumentCreate(BaseModel):
    doc_type: str  # PASSPORT, DRIVERS_LICENSE, NATIONAL_ID, SELFIE


class ValidationResult(BaseModel):
    is_valid: bool
    errors: List[str] = []
    extracted_name: Optional[str] = None
    extracted_dob: Optional[str] = None
    user_name: Optional[str] = None
    user_dob: Optional[str] = None


class DocumentResponse(BaseModel):
    id: str
    kyc_id: str
    doc_type: str
    file_path: str
    file_hash: str
    extracted_data: Optional[Dict[str, Any]] = None
    verified: bool
    uploaded_at: datetime
    validation_result: Optional[ValidationResult] = None  # Validation result if available
    
    class Config:
        from_attributes = True


class VerificationCreate(BaseModel):
    event_type: str
    details: Optional[Dict[str, Any]] = None
    performed_by: str
    tx_hash: Optional[str] = None


class VerificationResponse(BaseModel):
    id: str
    kyc_id: str
    event_type: str
    details: Optional[Dict[str, Any]] = None
    performed_by: str
    timestamp: datetime
    tx_hash: Optional[str] = None
    
    class Config:
        from_attributes = True


class KYCApplicationCreate(BaseModel):
    pass  # Created automatically when user uploads first document


class KYCApplicationUpdate(BaseModel):
    status: Optional[str] = None
    risk_score: Optional[float] = None
    face_match_score: Optional[float] = None
    reviewer_comment: Optional[str] = None


class KYCApplicationResponse(BaseModel):
    id: str
    user_id: str
    user_email: Optional[str] = None
    ukn: Optional[str] = None  # Unique KYC Number
    status: str
    risk_score: Optional[float] = None
    shap_explanation: Optional[List[ShapFeature]] = None
    face_match_score: Optional[float] = None
    reviewer_comment: Optional[str] = None
    user_details: Optional[Dict[str, Any]] = None  # User-entered details (name, DOB, etc.)
    blockchain_tx_hash: Optional[str] = None
    verified_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    documents: List[DocumentResponse] = []
    verifications: List[VerificationResponse] = []
    
    class Config:
        from_attributes = True


class RiskMetrics(BaseModel):
    total_applications: int
    auto_approved: int
    manual_reviews: int
    rejected: int
    avg_risk_score: float
    avg_processing_time: float


class AuditRecordResponse(BaseModel):
    id: str
    entity_type: str
    entity_id: str
    event_hash: str
    tx_hash: str
    timestamp: datetime
    details: Optional[Dict[str, Any]] = None
    
    class Config:
        from_attributes = True

